Whitelist in TS, GDB, 21.3.1.1-Final Mandrel Distribution (Java Version 17.0.2+8)

Wait for https://github.com/graalvm/mandrel/issues/355 ?
Workaround in TS / Whitelist for the time being?

GDB output differs from the expected one:

```
2022-03-24T18:38:05.1112648Z [ERROR]   DebugSymbolsTest.debugSymbolsSmokeGDB:137->carryOutGDBSession:386 There were errors in the GDB session. Note that commands in the session might depend on each other. Errors: 
2022-03-24T18:38:05.1115104Z Command 'info functions .*smoke.*' did not match the expected pattern '.*File debug_symbols_smoke/ClassA.java:.*java.lang.String \*debug_symbols_smoke.ClassA::toString\(void\).*File debug_symbols_smoke/Main.java:.*void debug_symbols_smoke.Main\$\$Lambda\$.*::accept\(java.lang.Object \*\).*void debug_symbols_smoke.Main::lambda\$thisIsTheEnd\$0\(java.io.ByteArrayOutputStream \*, debug_symbols_smoke.ClassA \*\).*void debug_symbols_smoke.Main::main\(java.lang.String\[\] \*\).*void debug_symbols_smoke.Main::thisIsTheEnd\(java.util.List \*\).*'.
2022-03-24T18:38:05.1124808Z Output was:
2022-03-24T18:38:05.1125181Z (gdb) All functions matching regular expression ".*smoke.*":
2022-03-24T18:38:05.1125390Z 
2022-03-24T18:38:05.1125662Z File debug_symbols_smoke/ClassA.java:
2022-03-24T18:38:05.1126465Z 	java.lang.String *debug_symbols_smoke.ClassA::toString(void);
2022-03-24T18:38:05.1126975Z 
2022-03-24T18:38:05.1127713Z File debug_symbols_smoke/Main$$Lambda$577/0x00000007c0582b28.java:
2022-03-24T18:38:05.1128120Z 	void debug_symbols_smoke.Main$$Lambda$577.0x00000007c0582b28::accept(java.lang.Object *);
2022-03-24T18:38:05.1128353Z 
2022-03-24T18:38:05.1128472Z File debug_symbols_smoke/Main.java:
2022-03-24T18:38:05.1128889Z 	void debug_symbols_smoke.Main::lambda$thisIsTheEnd$0(java.io.ByteArrayOutputStream *, debug_symbols_smoke.ClassA *);
2022-03-24T18:38:05.1129356Z 	void debug_symbols_smoke.Main::main(java.lang.String[] *);
2022-03-24T18:38:05.1134327Z 	void debug_symbols_smoke.Main::thisIsTheEnd(java.util.List *);
2022-03-24T18:38:05.1134670Z  ==> expected: <true> but was: <false>
```

--- a/testsuite/src/it/java/org/graalvm/tests/integration/DebugSymbolsTest.java
+++ b/testsuite/src/it/java/org/graalvm/tests/integration/DebugSymbolsTest.java
@@ -19,7 +19,6 @@
  */
 package org.graalvm.tests.integration;
 
-import org.graalvm.home.Version;
 import org.graalvm.tests.integration.utils.Apps;
 import org.graalvm.tests.integration.utils.ContainerNames;
 import org.graalvm.tests.integration.utils.GDBSession;
@@ -134,7 +133,7 @@ public class DebugSymbolsTest {
                                 3000, 500, TimeUnit.MILLISECONDS),
                         "GDB session did not start well. Check the names, paths... Content was: " + stringBuffer.toString());
 
-                carryOutGDBSession(stringBuffer, GDBSession.DEBUG_SYMBOLS_SMOKE, esvc, writer, report, UsedVersion.getVersion(false));
+                carryOutGDBSession(stringBuffer, GDBSession.DEBUG_SYMBOLS_SMOKE, esvc, writer, report, false);
 
                 writer.write("q\n");
                 writer.flush();
@@ -215,7 +214,7 @@ public class DebugSymbolsTest {
 
                 writer.flush();
 
-                carryOutGDBSession(stringBuffer, GDBSession.DEBUG_QUARKUS_FULL_MICROPROFILE, esvc, writer, report, UsedVersion.getVersion(false));
+                carryOutGDBSession(stringBuffer, GDBSession.DEBUG_QUARKUS_FULL_MICROPROFILE, esvc, writer, report, false);
 
                 writer.write("q\n");
                 writer.flush();
@@ -319,7 +318,7 @@ public class DebugSymbolsTest {
 
                 writer.write("set directories /work/sources\n");
                 writer.flush();
-                carryOutGDBSession(stringBuffer, GDBSession.DEBUG_QUARKUS_BUILDER_IMAGE_VERTX, esvc, writer, report, UsedVersion.getVersion(true));
+                carryOutGDBSession(stringBuffer, GDBSession.DEBUG_QUARKUS_BUILDER_IMAGE_VERTX, esvc, writer, report, true);
                 writer.write("q\n");
                 writer.flush();
             }
@@ -346,9 +345,9 @@ public class DebugSymbolsTest {
     }
 
     public static void carryOutGDBSession(StringBuffer stringBuffer, GDBSession gdbSession, ExecutorService esvc,
-                                          BufferedWriter writer, StringBuilder report, Version mandrelVersion) {
+                                          BufferedWriter writer, StringBuilder report, boolean inContainer) {
         final ConcurrentLinkedQueue<String> errorQueue = new ConcurrentLinkedQueue<>();
-        Stream.of(gdbSession.get(mandrelVersion)).forEach(cp -> {
+        Stream.of(gdbSession.get(inContainer)).forEach(cp -> {
                     stringBuffer.delete(0, stringBuffer.length());
                     try {
                         if (cp.c.startsWith("GOTO URL")) {

--- a/testsuite/src/it/java/org/graalvm/tests/integration/utils/GDBSession.java
+++ b/testsuite/src/it/java/org/graalvm/tests/integration/utils/GDBSession.java
@@ -20,6 +20,7 @@
 package org.graalvm.tests.integration.utils;
 
 import org.graalvm.home.Version;
+import org.graalvm.tests.integration.utils.versions.UsedVersion;
 
 import java.util.regex.Pattern;
 
@@ -31,14 +32,62 @@ import java.util.regex.Pattern;
 public enum GDBSession {
     NONE {
         @Override
-        public CP[] get(Version mandrelVersion) {
+        public CP[] get(boolean inContainer) {
             return new CP[]{};
         }
     },
     DEBUG_SYMBOLS_SMOKE {
         @Override
-        public CP[] get(Version mandrelVersion) {
-            if (mandrelVersion.compareTo(Version.create(21, 1, 0)) >= 0) {
+        public CP[] get(boolean inContainer) {
+            if (UsedVersion.jdkFeature(inContainer) == 17 && UsedVersion.jdkInterim(inContainer) == 0 && UsedVersion.jdkUpdate(inContainer) < 4) {
+                // workaround graalvm/mandrel#355
+                return new CP[]{
+                        SHOW_VERSION,
+                        new CP("run < ./test_data_small.txt\n",
+                                Pattern.compile(".*fdc7c50f390c145bc58a0bedbe5e6d2e35177ac73d12e2b23df149ce496a5572.*exited normally.*",
+                                        Pattern.DOTALL)),
+                        new CP("info functions .*smoke.*\n",
+                                Pattern.compile(
+                                        ".*File debug_symbols_smoke/ClassA.java:.*" +
+                                                "java.lang.String \\*debug_symbols_smoke.ClassA::toString\\(void\\).*" +
+                                                "File debug_symbols_smoke/Main\\$\\$Lambda.*.java:.*" +
+                                                "void debug_symbols_smoke.Main..Lambda..*::accept\\(java.lang.Object.*" +
+                                                "File debug_symbols_smoke/Main.java:.*" +
+                                                "void debug_symbols_smoke.Main::lambda\\$thisIsTheEnd\\$0\\(java.io.ByteArrayOutputStream \\*, debug_symbols_smoke.ClassA \\*\\).*" +
+                                                "void debug_symbols_smoke.Main::main\\(java.lang.String\\[\\] \\*\\).*" +
+                                                "void debug_symbols_smoke.Main::thisIsTheEnd\\(java.util.List \\*\\).*"
+                                        , Pattern.DOTALL)),
+                        new CP("break Main.java:70\n",
+                                Pattern.compile(".*Breakpoint 1 at .*: file debug_symbols_smoke/Main.java, line 70.*",
+                                        Pattern.DOTALL)),
+                        new CP("break Main.java:71\n",
+                                Pattern.compile(".*Breakpoint 2 at .*: file debug_symbols_smoke/Main.java, line 71.*",
+                                        Pattern.DOTALL)),
+                        new CP("break Main.java:76\n",
+                                Pattern.compile(".*Breakpoint 3 at .*: file debug_symbols_smoke/Main.java, line 76.*",
+                                        Pattern.DOTALL)),
+                        new CP("run < ./test_data_small.txt\n",
+                                Pattern.compile(".*Breakpoint 1, .*while \\(sc.hasNextLine\\(\\)\\).*",
+                                        Pattern.DOTALL)),
+                        new CP("c\n",
+                                Pattern.compile(".*Breakpoint 3, debug_symbols_smoke.Main::main.*at debug_symbols_smoke/Main.java:76.*String l = sc.nextLine\\(\\);.*",
+                                        Pattern.DOTALL)),
+                        new CP("c\n",
+                                Pattern.compile(".*Breakpoint 2, debug_symbols_smoke.Main::main.*at debug_symbols_smoke/Main.java:71.* if \\(myString != null.*",
+                                        Pattern.DOTALL)),
+                        new CP("c\n",
+                                Pattern.compile(".*Breakpoint 2, debug_symbols_smoke.Main::main.*at debug_symbols_smoke/Main.java:71.* if \\(myString != null.*",
+                                        Pattern.DOTALL)),
+                        new CP("d 2\n",
+                                Pattern.compile(".*", Pattern.DOTALL)),
+                        new CP("c\n",
+                                Pattern.compile(".*fdc7c50f390c145bc58a0bedbe5e6d2e35177ac73d12e2b23df149ce496a5572.*exited normally.*",
+                                        Pattern.DOTALL)),
+                        new CP("list ClassA.java:30\n",
+                                Pattern.compile(".*ClassA\\(int myNumber, String myString\\).*",
+                                        Pattern.DOTALL)),
+                };
+            } else if (UsedVersion.getVersion(inContainer).compareTo(Version.create(21, 1, 0)) >= 0) {
                 return new CP[]{
                         SHOW_VERSION,
                         new CP("run < ./test_data_small.txt\n",
@@ -136,7 +185,7 @@ public enum GDBSession {
     },
     DEBUG_QUARKUS_FULL_MICROPROFILE {
         @Override
-        public CP[] get(Version mandrelVersion) {
+        public CP[] get(boolean inContainer) {
             return new CP[]{
                     SHOW_VERSION,
                     new CP("b ConfigTestController.java:33\n",
@@ -157,7 +206,7 @@ public enum GDBSession {
     },
     DEBUG_QUARKUS_BUILDER_IMAGE_VERTX {
         @Override
-        public CP[] get(Version mandrelVersion) {
+        public CP[] get(boolean inContainer) {
             return new CP[]{
                     SHOW_VERSION,
                     new CP("b Fruit.java:48\n",
@@ -178,5 +227,5 @@ public enum GDBSession {
 
     private static final CP SHOW_VERSION = new CP("show version\n", Pattern.compile(".*", Pattern.DOTALL));
 
-    public abstract CP[] get(Version mandrelVersion);
+    public abstract CP[] get(boolean inContainer);
 }

